### ISSUE TITLE
Array module review followup: Should we remove sorted, reverse, find, and count from array type?

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#18089]
This is a followup item for the [Array module review](https://github.com/Cray/chapel-private/issues/2241).

One item we wanted further discussion on is:

These functions (remove sorted, reverse, find, and count) seem like nice utilities to have that operate on arrays but are not a fundamental part of the array type themselves.  Should we remove these functions?  Should we put them in some ArrayUtils or other module?

**Comment by mppf** [comment]
It makes sense to me to remove these and instead rely on other modules like Sort and Search to implement these.

**Comment by lydia-duncan** [comment]
I'm fine with having them no longer be defined on arrays themselves as long as their replacement is made available at the same time.

**Comment by bradcray** [comment]
I would definitely deprecate these, and am not sure I agree with Lydia's sentiment that replacements for them are strictly necessary.  Specifically:

* sorted() creates a temporary array and then iterates over it, yielding elements.  I'm not convinced this is better than simply requiring the user to call sort() on either the original array or a copy that they create (because it has hidden and potentially expensive costs; because forcing them to use an existing sort() routine exposes those costs).
* reverse() is specific to 1D arrays and can be written with a forall loop without too much difficulty (and, I suspect, it isn't used much in practice)
* find() is kind of goofy as currently written ("iterate over the array serially...") and would be better expressed as a reduction
* count() can trivially be written as a reduction

So I'd be more inclined to simply deprecate them (I'm not convinced they're things we want to be in the business of supporting), potentially giving hints as to how to rewrite them, or inviting users to pester us if they want to advocate for retaining any of them.

**Comment by mppf** [comment]
> find() is kind of goofy as currently written ("iterate over the array serially...") and would be better expressed as a reduction

That's interesting but I don't think we currently have a standard-module provided reduction that would do that, right? I guess you could write some combination of a forall expression over indices and `minloc` / `maxloc` ? It seems to me that if it were an operation we thought were common we might want to provide a reduction that does it.

**Comment by bradcray** [comment]
> It seems to me that if it were an operation we thought were common we might want to provide a reduction that does it.

That was my thinking as well.  I mostly think that providing a `find()` in a parallel language where the implementation is serial seems like a bad starting point, which is why I'm guessing nobody relies on it, and why I'm not keen on committing to interruption-free support of such a pattern.  That's why I was suggesting deprecating it, inviting users to argue that it's important, and only implement it if anyone does.

**Comment by stonea** [comment]
I've made a branch where I deprecate these functions and tried running paratest on it. These are the tests we're going to have to update once we realize this change:

```
[Error matching program output for arrays/diten/arrayCount]
[Error matching program output for arrays/diten/arrayFind]
[Error matching program output for arrays/diten/arrayReverse]
[Error matching program output for arrays/diten/enumArrayMethods]
[Error matching program output for arrays/diten/nestSortedIter]
[Error matching program output for arrays/userAPI/arrayOps2D (compopts: 1)]
[Error matching program output for arrays/userAPI/assocArray (execopts: 1)]
[Error matching program output for arrays/userAPI/assocArray (execopts: 2)]
[Error matching program output for arrays/userAPI/assocDistributedHashed (execopts: 1)]
[Error matching program output for arrays/userAPI/assocDistributedHashed (execopts: 2)]
[Error matching program output for arrays/userAPI/blockOps2D]
[Error matching program output for arrays/userAPI/blockOpsRankChange]
[Error matching program output for arrays/userAPI/blockOpsReindex]
[Error matching program output for arrays/userAPI/callExternArrTest]
[Error matching program output for arrays/userAPI/enumArray]
[Error matching program output for arrays/userAPI/rankChangeOps3to2D]
[Error matching program output for arrays/userAPI/reindexOps2D]
[Error matching program output for arrays/userAPI/sliceOps2D]
[Error matching program output for arrays/userAPI/unsignedOps]
[Error matching program output for associative/bharshbarg/arrays/largeLiteral]
[Error matching program output for associative/bharshbarg/domains/recordWithRange]
[Error matching program output for associative/bradc/assocRecord]
[Error matching program output for associative/bradc/removeResizeViaAssign]
[Error matching program output for associative/deitz/arrays/test_indefinite1]
[Error matching program output for associative/deitz/arrays/test_indefinite2]
[Error matching program output for associative/deitz/arrays/test_indefinite3]
[Error matching program output for associative/deitz/arrays/test_indefinite4]
[Error matching program output for associative/deitz/arrays/test_indefinite5]
[Error matching program output for associative/deitz/arrays/test_indefinite6]
[Error matching program output for associative/deitz/arrays/test_indefinite7]
[Error matching program output for associative/deitz/arrays/test_indefinite8]
[Error matching program output for associative/deitz/domains/test_indefinite1]
[Error matching program output for associative/deitz/domains/test_indefinite_domain_of_tuple]
[Error matching program output for associative/deitz/domains/test_indefinite_remove]
[Error matching program output for associative/deitz/domains/test_indefinite_remove2]
[Error matching program output for associative/diten/enumArrays]
[Error matching program output for associative/ferguson/assoc-sort-comparator]
[Error matching program output for associative/ferguson/plus-reduce-assoc]
[Error matching program output for associative/ferguson/plus-reduce-intent-assoc]
[Error matching program output for associative/waynew/indef6]
[Error matching program output for associative/waynew/indef7]
[Error matching program output for associative/waynew/indef8-fail1]
[Error matching program output for associative/waynew/indef8-fail2]
[Error matching program output for associative/waynew/indef8]
[Error matching program output for associative/waynew/indef8a]
[Error matching program output for associative/waynew/indef9]
[Error matching program output for associative/waynew/indef9a]
[Error matching program output for constrained-generics/basic/set2/resolve-implements-types]
[Error matching program output for distributions/bradc/assoc/userAssoc-array-domain-zipper]
[Error matching program output for distributions/bradc/assoc/userAssoc-basic-domain]
[Error matching program output for distributions/bradc/assoc/userAssoc-dom-operations]
[Error matching program output for distributions/robust/associative/basic/domain_write]
[Error matching program output for distributions/robust/associative/basic/whole_domain_assign]
[Error matching program output for distributions/vass/testReplicatedVar1]
[Error matching program output for distributions/vass/testReplicatedVar2]
[Error matching program output for domains/operators/domainAdd]
[Error matching program output for domains/operators/domainSub]
[Error matching program output for domains/sungeun/assoc/clear]
[Error matching program output for domains/sungeun/assoc/equals_minus_1 (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/equals_minus_1 (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/equals_minus_2 (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/equals_minus_2 (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/equals_plus_1 (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/equals_plus_1 (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/equals_plus_2 (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/equals_plus_2 (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/forall (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/forall (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/index_not_in_domain_1 (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/index_not_in_domain_1 (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/index_not_in_domain_2 (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/index_not_in_domain_2 (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/minus_equals (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/minus_equals (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/plus_equals (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/plus_equals (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/plus_minus_equals (compopts: 1)]
[Error matching program output for domains/sungeun/assoc/plus_minus_equals (compopts: 2)]
[Error matching program output for domains/sungeun/assoc/stress (compopts: 1, execopts: 1)]
[Error matching program output for domains/sungeun/assoc/stress (compopts: 1, execopts: 2)]
[Error matching program output for domains/sungeun/assoc/stress (compopts: 2, execopts: 1)]
[Error matching program output for domains/sungeun/assoc/stress (compopts: 2, execopts: 2)]
[Error matching program output for domains/sungeun/assoc/stress.numthr (compopts: 1, execopts: 1)]
[Error matching program output for domains/sungeun/assoc/stress.numthr (compopts: 1, execopts: 2)]
[Error matching program output for domains/sungeun/assoc/stress.numthr (compopts: 2, execopts: 1)]
[Error matching program output for domains/sungeun/assoc/stress.numthr (compopts: 2, execopts: 2)]
[Error matching program output for domains/vass/arrays-of-domains]
[Error matching program output for execflags/sungeun/about (execopts: 1)]
[Error matching program output for execflags/sungeun/about (execopts: 2)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc-reduce (execopts: 1)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc-reduce (execopts: 2)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc-reduce (execopts: 3)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc-reduce (execopts: 4)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc-reduce (execopts: 5)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc (execopts: 1)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc (execopts: 2)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc (execopts: 3)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc (execopts: 4)]
[Error matching program output for exercises/duplicates/duplicates-alt-assoc (execopts: 5)]
[Error matching program output for functions/deitz/iterators/leader_follower/test_neg_stride_range_leader_follower]
[Error matching program output for io/ferguson/assoc-array-chpl]
[Error matching program output for io/ferguson/assoc-domains]
[Error matching program output for mason/env/mason-env]
[Error matching program output for mason/env/mason-registry]
[Error matching program output for mason/mason-external/libtomlc99/mason-external]
[Error matching program output for mason/publish/noFork]
[Error matching program output for release/examples/primers/associative]
[Error matching program output for release/examples/primers/learnChapelInYMinutes]
[Error matching program output for studies/590o/alaska/graph]
[Error matching program output for studies/labelprop/labelprop-tweets (execopts: 1)]
[Error matching program output for studies/labelprop/labelprop-tweets (execopts: 2)]
[Error matching program output for studies/labelprop/labelprop-tweets (execopts: 3)]
[Error matching program output for studies/ssca2/graphio/graphio]
[Error matching program output for types/enum/diten/enumDom]
[Error matching program output for users/engin/sparse_bulk_dist/sparseAdd1D]
[Error matching program output for users/engin/sparse_bulk_dist/sparseAdd2D (compopts: 1)]
[Error matching program output for users/engin/sparse_bulk_dist/sparseAdd2D (compopts: 2)]
[Error matching program output for users/engin/sparse_bulk_dist/sparseBulk1D]
[Error matching program output for users/engin/sparse_bulk_dist/sparseBulk2D (compopts: 1)]
[Error matching program output for users/engin/sparse_bulk_dist/sparseBulk2D (compopts: 2)]
[Error matching program output for users/shetag/assoc]
[Summary: #Successes = 12999 | #Failures = 121 | #Futures = 0]
```

**Comment by bradcray** [comment]
@stonea: That's more failures than I was guessing we'd see.  Do you happen to have the `.log` file around for this run so I could see the failures in more detail?  (e.g., I'm wondering whether there's a common code path or two that accounts for a large number of the failures).  For example, it appears that associative domains/arrays may rely on one of these routines and account for the lion's share, such that a small edit or two to that module could reduce the number of failures greatly.

**Comment by stonea** [comment]
I'm taking the approach now of commenting out one function at a time and then manually looking at the errors to see the patterns.

Starting out with `sorted` here's the failing tests:

<details>

```
arrays/userAPI/arrayOps2D
associative/bharshbarg/arrays/largeLiteral
associative/bharshbarg/arrays/opEquals
associative/diten/enumArrays
associative/ferguson/assoc-sort-comparator
associative/waynew/indef8-fail1
associative/waynew/indef8-fail2
associative/waynew/indef8a
distributions/robust/associative/basic/array_write
domains/vass/arrays-of-domains
y/packages/DistributedIters/checkDistributedIters
library/packages/DistributedIters/twoLocalesCoordinated
library/packages/SortedMap/update/testUpdate
y/packages/TOML/BurntSushi/Passing
y/packages/TOML/test/TomlTest
library/packages/TOML/test/UTomlTest
library/packages/TOML/test/checkParseLoop
library/packages/TOML/test/commentTest
library/packages/TOML/test/date
y/packages/TOML/test/fileTest
library/packages/TOML/test/newline
library/packages/TOML/test/stringTest
library/packages/TOML/test/testtime
y/packages/UnitTest/Launcher_Test
library/standard/Map/testExtend
library/standard/Map/testInit
library/standard/Map/testIterators
library/standard/Map/testSetAssignOps
library/standard/Map/testSetOps
library/standard/Map/testToArray
library/standard/Set/setOperatorOrderTest
mason/build/noDeps
mason/chplVersion/mason-chpl-version-build
chplVersion/mason-chpl-version-update
mason/env/mason-registry
mason/mason-example-with-opts/mason-example
mason/mason-example/mason-example
mason/mason-external/libtomlc99/mason-external
mason/mason-external/masonExternalRanges/mason-external-range
mason-modify/masonModifyTest
mason/mason-offline/publishOffline
mason/mason-test/mason-test
mason-update-toml-error/mason-chpl-version-update
mason/masonInit/MasonInitTest5
mason/masonInit/masonInitModifyToml
mason/masonInit/masonInitModule
mason/masonInit/masonInitPathTest
mason/masonInit/masonInitTest
mason/masonInit/masonInitTest2
mason/masonInit/masonInitTest3
mason/masonInit/masonInitTest4
mason/masonTestSome/mason-test-some
mason/masonTestSubString/mason-test-sub-string
mason/pkgconfig-tests/openssl
mason/pkgconfig-tests/zlib
mason/publish/create-registry/create-registry
mason/publish/badDry
mason/publish/badRegTest
mason/publish/packageName
mason/publish/publish
mason/publish/publishHelp
mason/publish/publishHelpShort
mason/run/mason-run
mason/search/badFileName/mason-search
mason/search/cacheTest/mason-search
mason/search/mason-search
mason/subdir-commands/mason-run
masonUpdateTest
release/examples/benchmarks/shootout/knucleotide
release/examples/spec/Data_Parallelism/promotes-default
studies/590o/alaska/graph
studies/shootout/k-nucleotide/benmcd/knucleotide-user-hash
studies/shootout/k-nucleotide/bradc/knucleotide-blc
studies/shootout/submitted/knucleotide3
studies/shootout/submitted/knucleotide4
types/records/ferguson/array/array-in-record
types/records/ferguson/array/array-in-record2
types/records/ferguson/array/array-in-record3
types/records/ferguson/array/array-literal
types/records/ferguson/array/iterate-array-break
types/records/ferguson/array/iterate-array-zip
types/records/ferguson/array/iterate-array-zip2
types/records/ferguson/array/iterate-array
types/records/ferguson/array/iterate-array2
types/records/ferguson/array/local-array-access-call
types/records/ferguson/array/local-array-assign-read
types/records/ferguson/array/local-array-forallexpr
types/records/ferguson/array/local-array-init-read
types/records/ferguson/array/local-array-init-typed-read
types/records/ferguson/array/local-array-store-assign
types/records/ferguson/array/local-array
types/records/ferguson/array/tuple-arrays
types/records/ferguson/array/tuple-unpack-arrays
types/records/ferguson/array/whole-array-assign
types/records/ferguson/call/call-expr-tmp-in
types/records/ferguson/call/call-expr-tmp-return
types/records/ferguson/call/call-expr-tmp
types/records/ferguson/call/call-in 
types/records/ferguson/call/call-inout
types/records/ferguson/call/call-out
types/records/ferguson/call/call-out2
types/records/ferguson/call/call-ref
types/records/ferguson/iterators/iterate-mod
types/records/ferguson/iterators/iterate-not-yielded
types/records/ferguson/iterators/iterate-twice-not-yielded
types/records/ferguson/iterators/iterate-twice
types/records/ferguson/iterators/iterate-twice2
types/records/ferguson/iterators/iterate
types/records/ferguson/iterators/iterate2
types/records/ferguson/leak-futures/record-in-record-generic
types/records/ferguson/leak-futures/return-ref-arg
types/records/ferguson/multilocale/coforall-on
types/records/ferguson/multilocale/on-const
types/records/ferguson/multilocale/on-tuple
types/records/ferguson/multilocale/on-tuple2
types/records/ferguson/multilocale/on-tuple3
types/records/ferguson/multilocale/on-tuple4
types/records/ferguson/multilocale/on
types/records/ferguson/multilocale/on2
types/records/ferguson/multilocale/on3
types/records/ferguson/multilocale/on4
types/records/ferguson/parallel/begin-copy
types/records/ferguson/parallel/begin
types/records/ferguson/parallel/coforall
types/records/ferguson/return/return-assign
types/records/ferguson/return/return-assign2
types/records/ferguson/return/return-conditional
types/records/ferguson/return/return-global
types/records/ferguson/return/return-in-arg
types/records/ferguson/return/return-init-typed
types/records/ferguson/return/return-init-typed2
types/records/ferguson/return/return-init
types/records/ferguson/return/return-init2
types/records/ferguson/return/return-ref-lvalue
types/records/ferguson/return/return-ref-rvalue
types/records/ferguson/return/return-return
types/records/ferguson/return/return-tuple
types/records/ferguson/return/return-tuple2
types/records/ferguson/return/return
types/records/ferguson/return/return2
types/records/ferguson/tuples/call-in-tuple
types/records/ferguson/tuples/call-inout-tuple
types/records/ferguson/tuples/call-out-tuple
types/records/ferguson/tuples/call-ref-tuple
types/records/ferguson/tuples/class-tuple-record-array
types/records/ferguson/tuples/class-tuple-record
types/records/ferguson/tuples/tuple-arg-identity
types/records/ferguson/tuples/tuple-copy-declared-type
types/records/ferguson/tuples/tuple-copy-fields-declared-type
types/records/ferguson/tuples/tuple-copy-fields
types/records/ferguson/tuples/tuple-copy
types/records/ferguson/tuples/tuple-unpack
types/records/ferguson/tuples/tuple
types/records/ferguson/variables/for
types/records/ferguson/variables/global-assign
types/records/ferguson/variables/global-init-assign-typed
types/records/ferguson/variables/global-init-assign
types/records/ferguson/variables/global
types/records/ferguson/variables/global2
types/records/ferguson/variables/inner-proc
types/records/ferguson/variables/inner-proc2
types/records/ferguson/variables/local-assign-if
types/records/ferguson/variables/local-assign
types/records/ferguson/variables/local-init-typed
types/records/ferguson/variables/local-init
types/records/ferguson/variables/local
types/records/ferguson/variables/record-in-record
types/records/ferguson/variables/tuple-in-record
types/records/ferguson/variables/tuple
````

</details>

Seems like an "easy" way to update the tests would be to just add an iterator like this at the 
bottom of each one:

```chpl
iter sortedArray(A) {
  use Sort;
  var copy = A;
  sort(copy);
  for ind in copy do yield ind;
}
```

And then do something like `sed/s/sort\((.*)\)/sortedArray\(\1\))/g` to rewrite each one (I haven't verified if this is the right regexp, but you get the idea).

I started manually updating the first couple of tests and hit upon some problems:

- We have a .sorted() iterator for domains, seems like we should either have sorted for both of these
  or none of these.
- We have special behavior for when sorting an associative array. Specifically we rewrite it to:
  array.dsiSorted(chpl_defaultComparator()) 
  
Before proceeding I kind of want to get feedback about if this makes us reconsider removing this method (or if we think adding a sorted iterator to the sort module makes sense).

**Comment by stonea** [comment]
## Removing `reverse`

For `reverse` the test fallout isn't too bad:

```
arrays/diten/arrayReverse
arrays/userAPI/arrayOps2D
constrained-generics/basic/set2/resolve-implements-types
```

Reconstructing `reverse()` manually isn't totally trivial as it works by updating the array in
place rather than cloning it, iterating through the original array backwards, and copying it in.
Here's what it looks like as a standalone function is 

```
proc reverse(ref A) {
const lo = A.domain.low,
      mid = A.domain.sizeAs(A.idxType) / 2,
      hi = A.domain.high;
for i in 0..#mid {
  A[lo + i] <=> A[hi - i];
}
```

Anyway I have a commit on my cloned repos where I update the tests here, I can turn this into a PR later:

https://github.com/stonea/chapel/commit/4a4310ced2b7d60dea177a9c09320a91afa09297

The only part I'm questioning has to do with `resolve-implements-types.chpl`.  I think the interface of this test just needs to use some arbitrary method on the array type so I changed it from `reverse` to `isEmpty`.

**Comment by stonea** [comment]
## Removing `find`

For `find` there's a moderate number of failing tests:

<details>

```
arrays/diten/arrayFind
arrays/userAPI/arrayOps2D
arrays/userAPI/assocArray
arrays/userAPI/assocDistributedHashed
arrays/userAPI/blockOps2D
arrays/userAPI/blockOpsRankChange
arrays/userAPI/blockOpsReindex
arrays/userAPI/callExternArrTest
arrays/userAPI/enumArray
arrays/userAPI/rankChangeOps3to2D
arrays/userAPI/reindexOps2D
arrays/userAPI/sliceOps2D
arrays/userAPI/unsignedOps
sparse/CS/multiplication/performance
users/engin/sparse_bulk_dist/sparseAdd1D
users/engin/sparse_bulk_dist/sparseAdd2D
users/engin/sparse_bulk_dist/sparseBulk1D
users/engin/sparse_bulk_dist/sparseBulk2D
```

</details>

@bradcray, when you mention replacing calls to `find()` with a reduction were you envisioning a reduce intent in a forall loop, something like this:

```chpl
proc find(ref A : [], const ref needle) {
  var foundAtIdx = max(A.eltType);
  forall i in A.domain with (min reduce foundAtIdx) do
    if(A[i] == needle) then foundAtIdx = i;
  return (foundAtIdx != max(int), foundAtIdx);
}
```

Or were you thinking we should create a new kind of reduction so that users could express it with something like this:

```chpl
x = findloc(needle) reduce zip(A, A.domain);
```

I'm running into difficulties trying to figure out how I could do that.  The closest I could come up with is something like this where I hard-code in the value we're searching for:

```chpl
class findloc: ReduceScanOp {
  type eltType;
  var needle = 30; // How can I not hard code this?
  var value = max(int);

  proc identity return max(int);
  proc accumulate(x) { accumulateOntoState(value, x); }

  proc accumulateOntoState(ref state, x : eltType) {
    if x[0] == needle then
      state = x[1];
  }

  proc accumulateOntoState(ref state, x : int) {
    state = min(x, state);
  }

  proc combine(x) { accumulateOntoState(value, x.value); }
  proc generate() return value;
  proc clone() return new unmanaged findloc(eltType=eltType);
}

var A : [1..5] int = [10, 20, 30, 40, 50];
writeln(findloc reduce zip(A, A.domain));
```

If I simply add an init method like this and update the clone method like this:

```chpl
proc init(type eltType, needle : eltType) {
  this.eltType = eltType;
  this.needle = needle;
}
```

I'll get these errors:

```
foo.chpl:29: error: unresolved call 'findloc.init(eltType=type 2*int(64))'
foo.chpl:6: note: this candidate did not match: findloc.init(type eltType, needle: eltType)
foo.chpl:29: note: because call does not supply enough arguments
foo.chpl:6: note: it is missing a value for formal 'needle'
```

So I'm not sure what the Chapeltastic way to do this is.

**Comment by bradcray** [comment]
> For reverse the test fallout isn't too bad:

When I saw you started with the sorted case, my first reaction was "That sounds like the hardest one to start with!"  :D

Focusing on `find()` for the time being:

Looking at `find()` with fresh eyes today, I don't find it as objectionable as the others.  Specifically, `sorted()` seems annoying because it brings `Sort` into every compile that uses an array (which is all of them) and therefore more appropriate to have live in the `Sort` module (as a tertiary method?  or just a standalone routine?).  `reverse()` seems annoying in its 1D-ness.  `find()` and `count()` merely feel superfluous to me, but not as inherently objectionable, and I admittedly have more parallel experience than many new users.  It's also notable that they're not 1D-specific.  It may be that the serial implementation was the most objectionable part to me, or that my unhappiness with the first few bled into `find()` and `count()` as well (?).  And there is something to be said for supporting similar routines on arrays as strings when appropriate.

In suggesting a reduction be used instead, I didn't mean to imply that we needed to come up with that reduction ourselves or even necessarily give users a replacement for `find()` (instead, if we were to get rid of it, I'd just deprecate it and then help anyone who comes asking for a `find`).

I think the following reduction is pretty close to implementing the body of `find()`:

```chapel
...maxloc reduce zip(A == needle, A.domain)...
```

where the A == needle should result in a (conceptual) array of booleans and (I think?) our `maxloc` guarantees returning the earliest index in the array?  (so, the first true or "1").  If this were the body of `find`, some minor work might need to be done to make the result of the reduction match the interface of `find()` (?).

Again, looking at this again today, I'm wondering whether, if find() had been written this way (and count as `+ reduce (A == needle)`), I wouldn't have been so likely to throw it under the bus above.  It might be worth asking @mppf since he seemed interested in getting rid of these before I weighed in.

If we were to remove find(), I think most of the tests are pretty easy to clean up:

Any `userAPI` tests that I wrote (which I think is most / all of them) are testing the user API for arrays/domains/whatever, so the "fix" for them would be to remove the calls to these routines from them since the routines will no longer be part of the user API and to update their .good files .  Similarly, `arrays/diten/arrayFind` looks like a test added when `find` was added, so should similarly be removed.

`performance.chpl` seems to be using `find()` as something of a heavyweight `contains()`?

I'm guessing the find() in `modules/dists/SparseBlockDist.chpl` is the cause of the last four failures and is commendable in its use of it since...  checking what Block does, it has a little homegrown utility to do the same thing (though it probably also predated `find()`:

```chapel
proc Block.chpl__locToLocIdx(loc: locale) {
  for locIdx in targetLocDom do
    if (targetLocales[locIdx] == loc) then
      return (true, locIdx);
  return (false, targetLocDom.first);
}
```

This feels like a common idiom ("Where is my locale in the set of targetLocales?") so ideally something we'd write once and reuse everywhere rather than re-implementing for each distribution.

Of course, if we were to keep find(), we should just have `Block` use it instead.

Back on sort(), rather than looking at just removing it, I'd see whether we can move it to the `Sort` module somehow (which is a package module, so arguably just kicking the question down the road... but it addresses my main objection to the method by only bringing in `Sort` when a user wants to do some sorting).  Ideally that would mean not changing the tests, or maybe just not changing them much (?).

Gotta run, hope I edited this into a reasonable form before timing out.

**Comment by stonea** [comment]
## Removing count

Count's implementation is a simple reduction so it would be pretty easy to "manually inline" this wherever it's used:

```chpl
proc count(val: this.eltType): int {
  return + reduce (this == val);
}
```

We get a moderate number of failures if I just remove the method (mostly under `userAPI`):

<details>

```
[Error matching compiler output for arrays/diten/arrayCount]
[Error matching compiler output for arrays/userAPI/arrayOps2D (compopts: 1)]
[Error matching compiler output for arrays/userAPI/arrayOps2D (compopts: 2)]
[Error matching compiler output for arrays/userAPI/arrayOps2D (compopts: 3)]
[Error matching compiler output for arrays/userAPI/arrayOps2D (compopts: 4)]
[Error matching compiler output for arrays/userAPI/assocArray]
[Error matching compiler output for arrays/userAPI/assocDistributedHashed]
[Error matching compiler output for arrays/userAPI/blockOps2D]
[Error matching compiler output for arrays/userAPI/blockOpsRankChange]
[Error matching compiler output for arrays/userAPI/blockOpsReindex]
[Error matching compiler output for arrays/userAPI/callExternArrTest]
[Error matching compiler output for arrays/userAPI/enumArray]
[Error matching compiler output for arrays/userAPI/rankChangeOps3to2D]
[Error matching compiler output for arrays/userAPI/reindexOps2D]
[Error matching compiler output for arrays/userAPI/sliceOps2D]
[Error matching compiler output for arrays/userAPI/unsignedOps]
[Error matching compiler output for mason/mason-external/libtomlc99/mason-external]
[Error matching compiler output for mason/mason-external/masonExternalRanges/mason-external-range]
```

</details>

**Comment by stonea** [comment]
I'll have to wait until post code freeze to merge anything but I have PRs for these changes ready to go:
 
* https://github.com/chapel-lang/chapel/pull/20025
* https://github.com/chapel-lang/chapel/pull/20062
* https://github.com/chapel-lang/chapel/pull/20077
* https://github.com/chapel-lang/chapel/pull/20078

**Comment by bradcray** [comment]
Seeing #20140 get merged makes me realize that my "Hmm, _should_ we really be removing `find()` (and possibly `count()`) rather than just improving their implementation...?" musings [above](https://github.com/chapel-lang/chapel/issues/18089#issuecomment-1145466289) never really generated any discussion or response.  I guess technically I didn't ask any questions, but if it wasn't clear... I gained some cold feet when thinking about what it would take to implement efficiently (both in the sense of "seems tricky for a user" and "if this is the trick, maybe we should just do it."

Ditto for `count()`, though to a much lesser extent (in part because it's simpler to write, in part because I think it's less frequently used).

`reverse()` and `find()` I don't have any qualms about since they're so 1D-specific.

**Comment by stonea** [comment]
Hi Brad, what do you suggest we do at this point?  Certainly I can add `find` and `count` back in, perhaps marking them as unstable.

My argument for keeping them removed would be:
 
- for fundamental datatypes like arrays it seems like we should aim for a slim interface with utility functions put in some other module/class
- count seems easily implementable as a reduction
- find seems less obvious to me, it probably deserves its own function, but again should that function be on the Array datatype itself?

Anyway here's what I think the most viable options are:
- Leave things as is with these functions deprecated
- Reintroduce `find` into the Array datatype
- Reintroduce both `find` and `count` into the Array datatype
- Put `find` (and I suppose `count` too) in some other module, say `ArrayUtils`

If we reintroduce `find` we have the option of:
- Using its old implementation, or
- Doing it using a reduction

If we do reintroduce any functions to the Array type then we'd have the option:
- To mark unstable or not?

**Comment by bradcray** [comment]
I'm not sure what we should do, mostly because I don't know whether the general sense on the team is "No, we should definitely be getting rid of all of these" or whether people were fond of some and not others, and I/we (by some definition of "we") railroaded them into rejecting them in the last module review where we discussed them.

I don't feel as though these routines need to be replaced or the PR reversed immediately or anything like that (deprecation isn't removal after all), but after our exchanges and fact-finding on this issue, I did come away feeling like it was less obvious what we should do, and therefore likely worthy of more discussion.

I don't think we should re-introduce find() with its old implementation in any case.

Part of me agrees with the "slim interface" concept, but part of me also thinks of arrays as being somewhat string-like at times and/or strings as being somewhat array-like at times, so when there are things like .find() that seem like they would make sense to support uniformly across types rather than using a built-in method in one case and a module-based standalone function in another.  I also worry a bit about creating an ArrayUtils module if it had only 1-2 things in it.

**Comment by bradcray** [comment]
As a starting point, @e-kayrakli works a lot with strings and arrays, so I'd be curious where he stood on the spectrum of arrays should definitely vs. definitely not support `.find()` (just to pick a starting case).

**Comment by stonea** [comment]
> I don't feel as though these routines need to be replaced or the PR reversed immediately or anything like that (deprecation isn't removal after all), but after our exchanges and fact-finding on this issue, I did come away feeling like it was less obvious what we should do, and therefore likely worthy of more discussion.

Fair enough, I'll bring this to discussion.  My only worry would be if we're left without a consensus before the next release,  things get left deprecated by default, and then later we decide to un-deprecate it -- I'd rather not make things so unpredictable to our end users.  Anyway I suppose if that becomes imminent I can flip to `unstable` but I expect we'll have converged on a decision by that time anyway.

**Comment by bradcray** [comment]
> My only worry would be if we're left without a consensus before the next release, things get left deprecated by default, and then later we decide to un-deprecate it -- I'd rather not make things so unpredictable to our end users.

That's why I brought it up as soon as I saw that this had been merged.  :)

**Comment by e-kayrakli** [comment]
> As a starting point, @e-kayrakli works a lot with strings and arrays, so I'd be curious where he stood on the spectrum of arrays should definitely vs. definitely not support .find() (just to pick a starting case).

I am on the "definitely not" camp.

I see your point about arrays and strings being closer to one another, but it is not enough to convince me that we need a `find` on arrays as a method. This is somewhat subjective, but I don't view strings and arrays to be *that* close. In the past, we removed list-like functionality from arrays due to various reasons and arrays are more list-like than string-like to me.

I agree with Michael that we have better places for all of these functions (Search and Sort modules)... except for `reverse`. I am not sure what to do with it. But I can comfortably call it "let's remove until someone asks for that functionality", without looking at tests for potential use cases.

One counter-argument to keeping them (I can't see it being brought up on a quick look) is that efficient implementation of some/all of these could be dependent on how the array is implemented. A method would be the most natural way of adding these functionalities without exposing too much about the array interface. But I think it is a good exercise for us as language designers to come up with strategies that can allow efficient implementations by third parties by means of nice array interface and efficient tools (like `reduce`). There's always a next application that requires a more esoteric function to be run on arrays efficiently, after all.

**Comment by stonea** [comment]
# Summary from module re-review breakout session

We discussed whether or not `find` and `count` should be deprecated on the `Array` module in the module review meeting today (7/7/2022).

In attendance: @e-kayrakli, @daviditen, @brandonneth, @mstrout, @ronawho, @ShreyasKhandekar, @DanilaFe

Other people who come to mind that may be interested include: @bradcray, @mppf, @lydia-duncan

Here's my summary of what we discussed. Please speak up if you think I'm missing something important, misremembering something, or have other comments.

## Concerns

### Adding additional functions to array can add additional requirements on the type stored

Michelle was concerned that these functions would require that datatypes stored in Arrays to support the `==` operator (since find and count rely on it).  However, this concern was alleviated by the fact that failing to implement `==` (on a datatype stored in an array) would lead to errors at compile (not runtime) and would only occur if the user actually calls `.find()` or `count()`.

### These functions are convenient

Others (I believe Daniel, Brandon, and Shreyas) voiced that the liked having the convenience of functions like this around (relating it to experiences in other languages).  There's a concern that it might not be obvious how to write these reductions (especially to new users) and it might be easier to read code that calls out to named functions rather than using the reductions directly.

### If we move to a package module (that needs to be explicitly installed by the user) that's even more inconvenient

Daniel also voiced a concern that if we put this in another module it would be especially inconvenient if this module weren't part of the standard library (and was required to be installed, perhaps by package manager).  Our current working proposal (I go into details about this later in this post) is to put `find` and `count` in the `sort` module. `sort` is a package module meaning it's possible in the future we might intend for users to have to explicitly install these package via Mason.  If that happens it seems possible (likely?) that the `Sort` module would be made a standard module, however.

### These functions add bloat

Counter to this others (Engin and me) were concerned that this could lead to a "slippery slope" of what to include and not include in Arrays.

### Consistency across datatypes

Elliot brought up the point that if the `find` and `count` function exist on other fundamental datatypes (such as `map` or `list`) we'd probably want to include them for consistency.  If we did this we might also want these methods to have a consistent interface in terms of return values and arguments.

Looking at documentation I do see `find` and `count` on `List` but not `Heap`, `Map`, or `Set`.  `Count` is also on `String` as is `Find` although the interface is difference (Array's count returns a tuple whereas String returns the value -1 if it can't find).  String also includes `rfind` which is not on any of these other datatypes.  @ronawho, given that I didn't look this up during the meeting would change your vote (if `find`/`count` continues to exist in the `List` datatype)?

## Voting results

I held a vote to see people wanted to move it available a separate package.  Here are the results:

```
  find / count should stay on Array datatype  (no votes)
  find / count should be deprecated (and we don't need to supply them today)  (2 votes)
  find / count should be removed from array datatype but we need to have them elsewhere  (6 votes)
```

One argument for moving them to a new package is that doing so would allow us to leave these functions unstable for Chapel 2.0 and thus could later change their interface (or bring them back to the Array datatype if we heard demand for it). This argument swayed the opinion of some of the folks who otherwise would have wanted them to stay on the Array datatype.

The two votes to deprecate said they would also be ok moving them to a separate module so I think this is the working consensus.

As far as where to put it:

Engin suggested adding this to the existing `sort` module. I didn't hear any objections to this.  There was also a question of if these should be tertiary or free functions.  I suggested we keep them as free functions (like others in the module) and didn't hear any objects to that.

## Working proposal

**The current working proposal is:**

  These functions should be reintroduced as free functions into the `sort` module.

**We should also:**

  Consider if we should remove `find` and `count` from List (pinging @engin on this since you did the module review for List).

**Comment by lydia-duncan** [comment]
Note that `count` shouldn't matter on `Set`, since I believe Sets can only store each value once?  Count would maybe be interesting on Map w.r.t. the values stored, but would probably not be at all cheap to computer

**Comment by e-kayrakli** [comment]
> Our current working proposal (I go into details about this later in this post) is to put find and count in the sort module. sort is a package module meaning it's possible in the future we might intend for users to have to explicitly install these package via Mason. If that happens it seems possible (likely?) that the Sort module would be made a standard module, however.

`Search` module, not `Sort`.

> Consider if we should remove find and count from List (pinging @e-kayrakli on this since you did the module review for List).

This is the part that can sway me towards having these in the array type a bit. I don't think we should remove them from the data structures. If anything, we should add them to others that have them missing if it has a sensible meaning for the specific structure. Even with that thinking, I am reluctant to keep these in the array interface.

> Note that count shouldn't matter on Set, since I believe Sets can only store each value once? Count would maybe be interesting on Map w.r.t. the values stored, but would probably not be at all cheap to computer

`std::set` has `count`, AFAIK. And returns 1 or 0. I believe this has more to do with the ability to swap data structures (I doubt C++ STL does a good job of that), or having familiar interfaces.

> String also includes rfind which is not on any of these other datatypes.

This is a good argument against trying to draw parallels between array and string/bytes interfaces. Another point is that `find`/`count` on string/bytes does substring counting and finding, as well. I think that justifies having them on those types, but I don't think any of us wants to do sub-array finding/counting on Chapel arrays as part of the built-in interface. And this also ties in to the "slippery slope" argument. If we have a `find` for an element of the array, why not an overload that takes another array as an argument and do a sub-array find?

**Comment by bradcray** [comment]
> Search module, not Sort.

That's a relief.  Reading the comments in order, I couldn't really think of what would cause me to look for these in Sort.  :)

Assuming we keep these in some form, I think my preference continues to be to define them on the array type itself (as primary or tertiary methods) rather than standalone functions, even if there is an asymmetry with the details of what `find()` does on string, bytes, or list, because I think it results in code that's more pleasing looking.  And I'd probably keep them in ChapelArray from the perspective of convenience (no need to rely on another module) and a sense that these few more functions feel not so different from supporting `.contains()` on ranges or domains, `count`/`find` on strings/bytes, or other simple membership-style queries.

Philosophically, I think arrays are a very powerful, core type in Chapel, and it's not obvious to me why they shouldn't support these queries.  Or to turn that argument around, why wouldn't we put all count/find capabilities into Search?

All that said, I could live with what the majority voted on above, it's just not what I would choose.

> One argument for moving them to a new package is that doing so would allow us to leave these functions unstable for Chapel 2.0

I don't think this quite follows.  We can definitely mark things as unstable within the internal / standard modules for Chapel 2.0 as well.  They don't have to be in a package module to be considered unstable.

**Comment by bradcray** [comment]
> (such as sort)

Just to make sure we're on the same page—are we really talking about the `Sort` module, or is it `Search` as Engin said above?  (I still don't see anything about count/find that feels sort-like, but search-like makes sense to me).

> I'd say despite that it's still worth minimizing the number of things marked unstable in a module we're trying to stabilize.

I sort of agree with this, but not completely.  In part in this case because we don't really expose ChapelArray as a module to the user even though it's implemented as one.  We could potentially substitute "everything in the Arrays chapter is stable" except that we've already agreed that sparse arrays are not on deck for stabilization for 2.0 due to lack of time, so even discounting equals() (which is admittedly a bit of an oddball), I think we're far from accomplishing that, at least without breaking sparse arrays into their own chapter (which we could also do).

But I mostly want to come back to a question that I didn't intend to be rhetorical.  Restating it:  Let's say that I read the notes from the meeting and thought "This is exactly the right path!  In fact, we should move `string.find()`, `string.rfind()`, `string.count()` to `Search` as well, and make them standalone routines rather than methods given that they're all forms of searching and would make the string type slimmer (and maybe also `startsWith` and `endsWith` since they're search-like? and/or `replace`, since find and replace are usually adjacent in application contexts?).  Would your and Engin's reactions be "Yes, we're all in for that as well!" or not?  And if not, why not?

I'll stop trying to equate strings and arrays since that didn't really fly for Engin, but I'm not seeing why we wouldn't either (a) put all find/count routines into a single module with a single call style or (b) define them as consistently as possible as methods on the types for which they make sense rather than defining some of them one way in one place and others in another way in another place.

I don't think the argument can be bloat:  To my count strings have ~50 documented methods on them to arrays' ~30.  And I think of arrays as being a far more core/crucial/unique type to Chapel than strings, so arguably deserving of appropriately designed features.

I also don't buy the slippery slope argument that if we had a single-element array.find(), we would need to have a multi-element array.find() for consistency.  We can just decide not to do that if we don't find it compelling / part of our vision for arrays, or decide to add it if we did.

[Note that I probably sound far more zealous about this than I probably really am, but it stood out for me that the three newest people to Chapel in the meeting—Daniel, Brandon, and Shreyas—found the current placement the most intuitive and productive, which gave me significant pause since arguably we're trying to appeal to new users.  And makes me want to argue their side to make sure they're not getting written off due to being newer to the project.]

**Comment by bradcray** [comment]
[Sorry in advance that this is so long]

Let me start with Andy's ranty bit and turn it into a question:  I find the design process about as frustrating as anybody, particularly since I tend to need to get involved in most of the design discussions whether or not I'm leading them, but it's also not clear to me how to improve it (and I hope I've been clear that I'm open to doing so if we have some good ideas).  E.g., it's pretty easy for me to imagine establishing a convention like "class types in the standard libraries should start with capital letters" but much harder for me to imagine delineating guidelines that would answer a question like "should a given type support a given feature by default?" without the need for discussion.  Since you have experience with other projects and groups that I don't, do you have insights into things we ought to be doing here to address your concern, but are not?

(I also agree with, and appreciate, Engin's point that part of the 2.0 effort is intended to establish better for ourselves what these guidelines are.  But even then, I'm unclear on how we would codify them to cover future design questions that might come up).

And then to add my own rant:  Part of my frustration on this issue is that I'd intended my [comment on June 2](https://github.com/chapel-lang/chapel/issues/18089#issuecomment-1145466289) to be read pretty clearly as "Whoa, wait, I'm getting cold feet on this... are we going too far?"  But then rather than generating further discussion or consensus-building, the next action I was aware of was the PR that did the deprecation.  Maybe I just didn't voice my concerns clearly enough to make it clear that they were actually concerns and not just me sitting around playing devil's advocate?  And I suppose it could be argued that I was being inefficient in the first place by second-guessing an intention that we'd discussed and agreed upon in a deep-dive, but I also think it's unrealistic to assume that we'll always make the right decision when something is discussed for the first time (to my memory) before investigation is done.

---

Returning to the technical discussion:

Trying to wrestle more with what makes me wonder whether we should be keeping rather than deprecating _array.find() and/or .count, here are some thoughts that occur to me, some of which are likely to be repeats:

* I don't think any of our built-in types should become kitchen-sinky or feel the need to support _everything_, but I also don't think they should feel extreme pressure to be completely minimalistic / bare bones.  Finding that line is obviously the hard part.
* Starting from the extremes, things that seem more structural seem like obvious inclusions: .size, .dim(), and friends feel like they should clearly belong on the array type rather than being available via a distinct module.  And on the other side of the spectrum, things that get more numerical or algorithmic or imparting a specific interpretation on an array, like performing an FFT or matrix solve or transpose feel clearly out of scope.
* For me, sort() remains in the "not supported by default" category because it feels more clearly on the algorithmic side of the equation.  It also has the benefit of having an obvious/pre-existing/non-trivial module to associate it with if/when we wanted to keep it.  It also shares the following concern:
* reverse() isn't quite as algorithmic, but seems very specific to 1D (and possibly dense?) arrays, which is a red flag for me in that I'd prefer most automatic features on arrays to work on most array types (where obvious exceptions would be cases that are specific to the structural nature of only a given array type, like a "number of nonzeroes" query on a sparse array).  
* find() and count() are definitely in a grey-er area for me, as this discussion has proven.  On one hand, they're not so intrinsically associated with the structure of the array that not having them impairs your use of the array.  Yet they are arguably a bit related to the structure in the sense of "Does this collection store this value at all?  Does it store multiple of them?"  They also don't seem like particularly deep concepts and like the sort of thing you'd ideally want on any given collection type.  They seem clearly rank-neutral which lets them pass that test.
* Of the two, I'm not entirely sure why, but I lean more toward including find() by default whereas count() I don't feel as strongly about.  But I'm not trying to throw count() under the bus, just pointing that observation out if anybody wanted to fight for one but not both.

I suppose these bullets could be interpreted as starting to develop a (personal) philosophy that could serve as future guidelines, but in many respects they seem pretty specific to arrays.  For example, I'm not sure that from them we could make better decisions about what locales or bigints or heaps should support.

I appreciate that you agree that treating these methods as similarly as possible on arrays and strings makes sense [Andy, at least.  I guess Engin didn't seem convinced]. If we agreed as a team to do that, it would make this decision much easier for me because I wouldn't want to have to use a module like that to get find() on strings, nor would I want find() on strings to turn into a standalone, rather than method-based routine.  So it would break my personal uncertainty on _array.find() by raising the spectre of making the string case so much less attractive that I'd definitely vote hard for "Let's just agree to support .find on arrays and strings and lists and other appropriate collections by default then."  Maybe this suggests a next step here:  To see whether the team:
* thinks these should be treated similarly across types or not (and if not, by what rationale)
* has a preference between keeping them on the types vs. moving them into a helper module if "similarly" wins the prior vote

(another possible next step is to make it obvious that I'm in a distinct and tiny minority and wasting everyone's time by belaboring this conversation... but the readout from the discussion and Andy's reaction to the string suggestion suggests to me that it's not quite that bleak).

Having typed all this, I'm realizing that part of my reason for not wanting a `Find` or `Search` module that defines `list.find`, `array.find`, `string.find`, etc. is that I feel as though for many of these types, an efficient implementation would likely want to interact with the type's internal storage/structure rather than being implemented through its public interfaces (I think it's definitely the case for strings, somewhat the case for maps, much less the case for arrays and possibly bytes).  Ironically, it was the perceived simplicity of implementing these using public interfaces on arrays that motivated us to consider deprecating them in the first place, and I'm not quite sure what to make of that other than "Maybe arrays are so structurally simple and/or tuned via their public interfaces that it makes more sense of them than others?"  So maybe a principle here is "If the routine is likely to want to access the type's internal structure to get an efficient/parallel implementation, it should be part of the type?"

There also seems to be a challenge to "put all find() methods into a Find/Search module" philosophy in that it's not extendsible to new collection types created by a user.

One other observation would be that if we had interfaces implemented, perhaps we'd think of these things in terms of being part of a "Findable()" interface and then think about whether a given type does or does not fulfill that interface.  And then define the set of possible interfaces that our built-in and/or standard types will support and those that are definitely off the table (e.g., Sortable).

I also appreciate your introspection about why you're inclined to move them out.  I think my rebuttal there would be that while, as Chapel implementers, we see these things as classes or objects because we're building them, to the user they're more like intrinsic types, so it gets into a question of whether it's better/nicer/more convenient to have feature-rich built-in types or feature-poor types.  Put another way, for this question I think we should focus most on what's attractive/useful/productive to users rather than what's easiest for us when defining the built-in types from a SW engineering perspective.  The fact that (I think) these routines have fairly straightforward implementations also makes the code creep issue seem less problematic to me (whereas `sorted` required dragging in the whole `Sort` module which seems like an obvious red flag).

> if we were to jam these methods in another internal module like ChapelArrayUtils, which is public used from ChapelArray

I don't like this approach better than the extremes.  As a user it means that I still get the feature by default, but don't see it documented alongside other array capabilities when I'm reading the language spec (or, if I do, then it completely doesn't matter to me as a user so seems no better than simply putting it in ChapelArray).  And in either case, as a developer, I'd similarly rather see all of the "available by default" methods on arrays defined in a single place rather than learning which file to look in for which features.  If we did unify the documentation, it's another hassle to wrestle with making happen; and if we didn't, it's a slightly clunky module in a list that I'd prefer not to have/start.

**Comment by e-kayrakli** [comment]
Your comments generally resonate with me and as I was mentioning before, I am much closer to the fence (not to imply that you should try harder to convince me, I just need to mull on it. And I wouldn't hold progress on this just for my sake.)

> Having typed all this, I'm realizing that part of my reason for not wanting a Find or Search module that defines list.find, array.find, string.find, etc. is that I feel as though for many of these types, an efficient implementation would likely want to interact with the type's internal storage/structure rather than being implemented through its public interfaces (I think it's definitely the case for strings, somewhat the case for maps, much less the case for arrays and possibly bytes). Ironically, it was the perceived simplicity of implementing these using public interfaces on arrays that motivated us to consider deprecating them in the first place, and I'm not quite sure what to make of that other than "Maybe arrays are so structurally simple and/or tuned via their public interfaces that it makes more sense of them than others?" So maybe a principle here is "If the routine is likely to want to access the type's internal structure to get an efficient/parallel implementation, it should be part of the type?"

This implies that we should also extend the optional `doi*` interface to include `doiFind`, `doiCount` and whatever we may want to add to the array interface. Which, when I think about it, is just a natural part of having these functions as part of the array interface.

**Comment by stonea** [comment]
Like Brad, I'll split this post into separate sections on the higher level meta-discussion and this more specific technical issue too.

## Meta-discussion:

In regards to the meta-discussion, let me know if you feel this is something best taken offline or tabled until we're going through our process review.

I think if we were working physically in an office this kind of meta-discussiony thing is something I'd be bringing up in more impromptu hallway conversations / lunch topics and the like.  It's less clear to me what the "right" way to do this in a remote working environment.  I'd rather error on the side of being communicative than not, but, ya' know', I'm also interested in not being critical in an unhelpful way or making an ass of myself :-), so let me know if you think bringing this up on a Github issue (on the private tracker of course) is going too "off the rails" so to speak.


> it's pretty easy for me to imagine establishing a convention like "class types in the standard libraries should start with capital letters" but much harder for me to imagine delineating guidelines that would answer a question like "should a given type support a given feature by default?" without the need for discussion

Guidelines are meant to provide guidance not unambiguously answer every possible question that could come up without discussion.  My hope is it would lead to more consistency, serve as something we can point to when asked "why did you choose x rather than y", and serve as a tie breaker when it comes to decisions that seem more "a matter of style".

Regarding "should a given type support a given feature by default?", I imagine we could have something like:

"when it comes to fundamental datatypes (arrays, domains, strings, etc.) we want the language to be feature rich and for tools to be accessible and discoverable.  We want users to feel that Chapel is friendly and easy-to-use, they shouldn't have to go digging to figure out where something is and we shouldn't be afraid to have multiple ways of accomplishing the same thing."

or:

"when it comes to fundamental datatypes we want interfaces to be simple, think the "Unix philosophy": do one thing and do it well. This makes it faster to compile (once we have incremental compilation that is), makes it easier to document and describe these classes, leads to more modularity, and makes these classes more robust to change"

or (if you want to try and strike a balance):

"We want fundamental datatypes to be simple but not to the point of being bare bones. We don't want bloat but having a small set of commonly used tools handy in the datatype itself is not a bad thing. How do we decide what to keep and what to cut? It's a tricky judgement call to make but it can help to ask questions like: is this commonly used, do we see it in benchmarks we've implemented or user codes like Arkouda and Champs? Do most other languages have it in their API? etc.

To give a sense of what we're aiming for here are some examples of past decisions we've made:

- things that are more structural like: .size, .dim(), etc. (on arrays and domains) belong as methods
- things that are more algorithmic like: sort() belong in their own modules
- ... "


>  Since you have experience with other projects and groups that I don't, do you have insights into things we ought to be doing here to address your concern, but are not?

I wish I could say "Oh yeah, I have direct experience doing this and here's what I've seen work", but I do think there is something unique about our effort to standardize our standard modules that makes it different than design discussions I've had in the past.

If you asked me: "well, even if you weren't directly involved, what have you seen before?"  I guess I'd say decisions involving changing the core of the language involved a lot more hierarchy and process, things more niche or experimental are a lot more agile driven largely by small groups, hallway conversations, and a lot of back and forth with code changes and design discussions.  If you want me to be less vague, it's something we could talk about offline.

If you're asking for a concrete decision today, I'd suggest:

* we develop a set of language design guidelines as we go
* we introduce some process to evaluate past decisions we've made as we modify and add to the guidelines
  * when we inevitably find inconsistencies, we we either reevaluate the past decision, update the guidelines, or decide that this really is an exceptional case.

I have to admit I somewhat reticent to suggest this since it sounds like a bunch of work, it's not 100% obvious it would help, and I'm opening myself up to the response "hey, Andy that sounds like a great idea, why don't you lead that effort :-)".

If I can't get the whole "let's create guidelines as we go" thing to fly my next suggestion would be once we're further along stabilizing things we really should have someone do a pass or two going through the modules collectively and try and find inconsistencies in the design.

More generally, and hopefully I don't get branded a heretic for saying this :-), but I also don't feel we should feel beholden to arriving at consensus for every decision.  Sometimes it would be more efficient to timebox decisions and go with the majority, or have leadership play the role of tie-breaker, or have a designated "design Tsar" who serves as a tie-breaker, or have the group agree to disagree and commit on the majority vote, or delegate a decision to a particularly interested stakeholder (or small group of stakeholders).

I'm totally open to hearing that I'm wrong here BTW.  I think there are benefits to a consensus driven approach, but they do come at the cost of efficiency and there's a slew of other pathologies that can emerge if not actively tempered against (e.g. inconsistent design, the loudest voices overshadowing others, relatively trivial but understandable issues getting more attention than deeper ones, people feeling less involved/responsible for any given issue, etc.)

How to temper against it? That's the million dollar question I suppose. If you want my answer it'd be trying to identify these issues and adding processes to address them or to having particularly involved leadership guiding the whole thing (which is maybe what we're doing now: you mentioned earlier that you were frustrated that you "tend to need to get involved in most of the design discussions whether or not [you're] leading them")

FWIW by-in-large I think the Chapel 2.0 effort is working to improve things so maybe that in itself is evidence that the approach we're taking is the right one.

> And then to add my own rant: Part of my frustration on this issue is that I'd intended my comment on June 2 to be read pretty clearly as "Whoa, wait, I'm getting cold feet on this... are we going too far?" But then rather than generating further discussion or consensus-building, the next action I was aware of was the PR that did the deprecation.

My reading of the comment was that you became less convinced that we should deprecate these functions but not necessarily that you felt deprecating them would be a mistake in itself.

I think your biggest concern was with `sort`, suggesting I "see whether we can move it to the Sort module somehow", which my PR effectively does (`Sort` already has a `sorted` function, which I adapted to so it would work anywhere the "old" `.sort()` does).  Given that we otherwise had consensus I didn't view this as meaning "hey, I think this decision needs to be reconsidered".  PRs go through their own review process and we can always un-deprecate things so having a code change be the next move seemed pretty low risk to me.

In hindsight I should have asked for clarification and I definitely don't want anyone to feel I'm ignoring their concerns so I apologize for that.

---

## More focused technical discussion:

Ok so back to talking about this specific Github issue:

> Starting from the extremes, things that seem more structural seem like obvious inclusions: .size, .dim(), and friends feel like they should clearly belong on the array type rather than being available via a distinct module

Agreed. Querying size seems like asking a fundamental question about the array itself that almost any large Chapel program is going to have to do several times.  I suppose one could make the argument that this is really a property of the domain but querying size is such a common thing to do that forcing users to go through an extra step of querying domain seems needlessly.


> reverse() isn't quite as algorithmic, but seems very specific to 1D (and possibly dense?) arrays, which is a red flag for me in that I'd prefer most automatic features on arrays to work on most array types (where obvious exceptions would be cases that are specific to the structural nature of only a given array type, like a "number of nonzeroes" query on a sparse array).

That convinces me that in the short term we should deprecate reverse (or at least mark it unstable).  I don't think `reverse()` necessarily has to be a 1D operation: exactly what it should mean for N dimensions is questionable (I'd say whatever arrangement would cause the iteration order to reverse after applying it).


> find() and count() are definitely in a grey-er area for me, as this discussion has proven. On one hand, they're not so intrinsically associated with the structure of the array that not having them impairs your use of the array. Yet they are arguably a bit related to the structure in the sense of "Does this collection store this value at all? Does it store multiple of them?" They also don't seem like particularly deep concepts and like the sort of thing you'd ideally want on any given collection type. They seem clearly rank-neutral which lets them pass that test.

I kind of like leaving count out just because `+ reduce (A == needle)` seems like such an elegant way of expressing the computation to me.  I'd advocate whatever decision we make re: count for Arrays we do the same for Strings.

The reduction for find `maxloc reduce zip(A == needle, A.domain)` seems a lot less obvious so I think providing an implementation somewhere makes sense.  Our previous consensus was putting it in the `search` module and if you asked me to decide that's still where I'd put it but if the winds have changed on this I'm willing to concede.

One advantage I see to making it a free function in the `search` module rather than a method is that opens up the possibility for us to make this function generic in the sense that it would work on any new datatype a user creates (assuming it fulfills some concept such as being iterable).  Of course we could have both a generic version in `search` and have methods on our datatypes (which might just call out to the version in `search` anyway) but if users get in the habit of using using a method version of `.find` they'll be less likely do discover the free function version.


> Maybe this suggests a next step here: To see whether the team:
> thinks these should be treated similarly across types or not (and if not, by what rationale)
> has a preference between keeping them on the types vs. moving them into a helper module if "similarly" wins the prior vote

Sure, I can try and drive this forward. My inclination at this point is to try and keep the conversation online rather than "in person" (so to speak), I can try and summarize do the polling thing.

Just to repeat my sense of where things stand (let me know if you disagree):

As I see it right now it seems like we have consensus that `reverse` should be removed.  It also seems like folks are happy with deprecating `.sort()` so long as the `sorted` iterator in the `Sort` module.

Of course people are always free to comment, but I think the polls I'd bring up are:

- If you were the only decision maker what would you do with `find`: (1) make it a methods in Array, (2) make it free functions in the `Search` module, or (3) remove it altogether

- If you were the only decision maker what would you do with `count`: (1) make it a method in Array, (2) make it free functions in the `Search` module, or (3) remove it altogether

- If we were to make `find` a free function in the `Search` module instead of a method on Array (for Chapel 2.0) would you feel that we were making a mistake?  yes, no, or I'm not sure

- If we were to make `count` a free function in the `Search` module instead of a method on Array (for Chapel 2.0) would you feel that we were making a mistake?  yes, no, or I'm not sure

If we get everyone to vote the same on the last two polls I'd consider that consensus (not unanimity, but I don't think we aiming for unanimity, right?) Do you feel this is the right approach to take or do you think I'm being too leading with these last two poll questions?


> There also seems to be a challenge to "put all find() methods into a Find/Search module" philosophy in that it's not extendsible to new collection types created by a user.

I guess I just made the opposite argument above. It seems like for things like `find` and `count` it's pretty easy to write a generic implementation that works on any iterable datatype.  Sure that generic implementation may not be the most efficient but users can always override that generic function if/when efficiency becomes a concern.

I guess your argument is in order to write that efficient implementation users would need to expose the private members of their class? I don't have a good sense of if that would be the case more often than not. I suppose if Chapel were C++ I'd say: write a private `findImpl()` method for your datatype, have the public `find()` function `friend` that method and call out to it. I'm not sure if we want to have a Chapel equivalent to the `friend` keyword though.


> Put another way, for this question I think we should focus most on what's attractive/useful/productive to users rather than what's easiest for us when defining the built-in types from a SW engineering perspective

I'm not sure these things are separable as you're suggesting.  Of course I agree we should be user focused.  But our users are programmers who are going to use and extend these datatypes too and given that it seems like whatever arguments one can make for what's right from a "SW engineering perspective" would be applicable to our users as well.  Of course what's considered right form "SW engineering perspective" is a hugely debatable topic in itself.

**Comment by stonea** [comment]
While looking at the docs @arezaii noticed that my PR didn't include a workaround for `reverse` in the deprecation message. Assuming we decide to keep this deprecated should probably update it to include a workaround (I think just add a loop through the array that goes in reverse order).

All the other removed functions have workarounds in their messages (as can be seen in this `.good` file for this test: https://github.com/chapel-lang/chapel/blob/main/test/deprecated/deprecatedArrayFunctions.good)

**Comment by bradcray** [comment]
I don't know that putting a workaround in the deprecation message is strictly required.  If a user is relying on it and doesn't know where to turn, I'm sure we'll hear from them (and the deprecation message would disappear in a release or so anyway).

**Comment by bradcray** [comment]
From Engin's last comment:

> This implies that we should also extend the optional doi* interface to include doiFind, doiCount and whatever we may want to add to the array interface.

I think we definitely could do that, but don't know that we need to do it, or should do it before we do have a clear need.  Specifically, I think find and count can be written plenty efficiently using the current dsi interface since the array is already telling us things like how to iterate over it optimally, and that's all that's required to do these ops well.  So while I wouldn't say "never!", and the doi interface can always be extended, I also don't see a significant call for it at present.

Rethinking the comment that led to this, however, maybe I'm wrong that we could write a faster find/count on strings using the internal implementation than a user/standalone library routine code?  We might just need to do a zippered iteration over the strings codepoints and 0..<size, whereas I may have only been thinking of using just the latter and random access when I wrote that (?).

Switching to an Andy comment I never replied to:

> I kind of like leaving count out just because + reduce (A == needle) seems like such an elegant way of expressing the computation to me.

I agree it's elegant, and I definitely find it attractive, but it does require a slightly greater level of expertise in Chapel to write and know to write, relying on both promotion and reductions.  While the deprecation warning is there, users will have their hand held, but I don't think we expect to keep the deprecation warning for all time.  In contrast `.count()` is something that any new-ish programmer might go looking for and know how to use.

> I'd advocate whatever decision we make re: count for Arrays we do the same for Strings.

I like that, but strings don't have a similarly elegant and concise way of writing count, though, do they?  This definitely relates to why I'm advocating for a count method.

> Our previous consensus was putting it in the search module

"Our" here refers to a subgroup of people who met to discuss this, not the team as a whole, though, right?

> Of course people are always free to comment, but I think the polls I'd bring up are:

I think your poll questions need to bring strings into the mix given that treating them symmetrically seemed to have resonance for both you and Engin.  For example:

Q: should `find()` be:
(a) a method on strings but a standalone routine in the `Search` module for arrays ("the previous consensus")
(b) a method on strings but not supported for arrays at all ("the status quo")
(c) a method on both strings and arrays, defined where the types are
(d) a standalone routine for both strings and arrays defined in the `Search` module
(e) something else
?

I also feel like in your list of four questions, the second pair are redundant with the first.  I.e., if you want to see how strongly people feel about an option, have them rate each on a scale of 1-7 or use approval voting rather than asking which they want and then asking how much they do or don't like the one you're advocating for.

> Sure that generic implementation may not be the most efficient but users can always override that generic function if/when efficiency becomes a concern.
> ...
> I guess your argument is in order to write that efficient implementation users would need to expose the private members of their class?

Right, after your first paragraph, I'd written:  Users can't really override the generic function without adding a method to the type, can they?  So type authors could add a method to their type to provide more efficient implementations, but clients of a type can't?

**Comment by stonea** [comment]
Hi @bradcray,

We're approaching our code freeze deadline and given that we don't have consensus on having  these functions deprecated I'd like to move these functions to be 'unstable' instead.

Alternatively I could just reverse the deprecation as a stopgap measure if you feel that would be better.

This isn't to say I'll be closing this issue; I am planning on conducting a poll to see how people feel about this in relation to strings as you suggest. I just don't want to leave these deprecated as part of the release unless we're sure that's how they're going to stay.

**Comment by bradcray** [comment]
> I'd like to move these functions to be 'unstable' instead.
> 
> Alternatively I could just reverse the deprecation as a stopgap measure if you feel that would be better.

Either of those would be fine with me and preferable to deprecating them at this point.

>  I am planning on conducting a poll to see how people feel about this in relation to strings as you suggest.

:+1:

**Comment by stonea** [comment]
On 10/11/2022 we had a re-review meeting to discuss this issue. Coming out of it we agreed to the following:

- We're going to deprecate: `sorted` and `reverse` on Array: nobody spoke up in favor of keeping these.

- We're going to keep (as unstable): `find` and `count` in the Array datatype

- There are inconsistencies in the interface for `find` and `count` across the `array`, `bytes`, `string`, and `list` datatypes. So we'll leave `find` and `count` as unstable for now with the goal of reevaluating these interfaces and seeing if we want to make them more consistent. The intent is not to have these left as unstable past 2.0: the functionality will be there, and will be accessible as a method on the array datatype, but the exact interface may change between now and then.

**Comment by bradcray** [comment]
While working on my recent reverse-complement implementation in the evenings (#21329), I was shocked at how hard it was to write a Chapel-level `find()` routine that could compete with C's `memchr()` (for arrays of 8-bit values), and what a big impact this seemingly innocuous input post-processing step had on the overall execution time of what I think of as a fairly complex computation.   For example, when piping the output to `/dev/null`, [Elliot saw improvements](https://github.com/chapel-lang/chapel/pull/21329#issuecomment-1376056135) from ~1.03–1.09s for the best Chapel-based implementations to ~0.82s for a memchr-based version.  And in nightly testing, when capturing the output, last night we saw a difference of ~1.94s to ~1.78s (comparing gcc8 transliteration to gcc8 translit (memchr) on [this graph](https://chapel-lang.org/perf/chapcs/?startdate=2022/10/19&enddate=today&graphs=reversecomplementshootoutbenchmark)):

These experiences put me more firmly in the "array's should support `find()` out of the box" camp (not that we're debating that anymore), as well as that their interfaces should be more similar to that of `.find` on `string`/`bytes` (which is to say, should return the index, where I'd have it return an OOB index of some sort when the find fails).

When Engin originally mentioned the `doiFind()` addition to the domain map interface, I was skeptical that it was necessary, but now I'm much more open to it, having failed to write an implementation that could compete with `memchr()` (suggesting that a great implementation might use `memchr()` on each locale for a distributed array, or in parallel on each locale to be even faster).  As an implementation path, my thought is that we should:

* getting the interface right
* implementing a catch-all solution that works for all rectangular arrays x eltTypes
* implementing an optimized version for DR arrays of int(8)/uint(8) that uses memchr()
  * also optimize bytes/ASCII-string find() to use memchr()
* optimizing other cases as motivating uses come up

It would be cool to add a revcomp version that uses array.find() and then see it improve over time (but it'd be more motivating to do this after stabilizing the interface).

**Comment by bradcray** [comment]
After running into a desire for array.find() in my recent CLBG push, I prototyped a version of the new interface this weekend and kicked off a design issue about its interface here: https://github.com/chapel-lang/chapel/issues/21423

**Comment by stonea** [comment]
I think I can close this issue as we've implemented the decisions listed here:

https://github.com/chapel-lang/chapel/issues/18089#issuecomment-1276500706

As well as updated the interface of `find` in this PR: https://github.com/chapel-lang/chapel/pull/21429

**Comment by bradcray** [comment]
@stonea : Do we have an issue about what needs to be done to make `count()` stable?  (that's the main reason I didn't have #21429 close this issue)

**Comment by stonea** [comment]
> @stonea : Do we have an issue about what needs to be done to make count() stable? (that's the main reason I didn't have https://github.com/chapel-lang/chapel/pull/21429 close this issue)

Good point. I've created an issue for that here: https://github.com/chapel-lang/chapel/issues/21730

**Comment by bradcray** [comment]
Thank you!
